### PR TITLE
[WIP] Adding FulfillmentOrder, FulfillmentRequest, CancellationRequest, AssignedFulfillmentOrder resources/filters

### PR DESCRIPTION
**Pull request by karmakaze**
Issue [#211513](https://github.com/Shopify/shopify/issues/211513) make shopify_api FO compatible

Add `/orders/:order_id/fulfillment_orders` resource

Reviewers please carefully check the `move` endpoint test in particular for form of request query/body and response.

### CODE REVIEW

File: test/fixtures/fulfillment_order.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      "delivery_category": null,
+      "fulfillment_order_line_items": [
+         {
+            "id": 519788021,
+            "shop_id": 690933842,
+            "fulfillment_order_id": 519788021,
+            "line_item_id": 466157049,
+            "variant_id":  null,
+            "inventory_item_id":  null,
+            "inventory_commitment_id":  null,
```
Comment (justinkrol) [inline test/fixtures/fulfillment_order.json]: I think it would make sense for these (`variant_id`, `inventory_item_id`, `inventory_commitment_id`) to be set. In the real system, we shouldn't have `null` values here.

Comment (karmakaze) [reply test/fixtures/fulfillment_order.json]: Ok, I'll find and add some values.

File: test/fixtures/fulfillment_order.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      "delivery_category": null,
+      "fulfillment_order_line_items": [
+         {
+            "id": 519788021,
+            "shop_id": 690933842,
+            "fulfillment_order_id": 519788021,
+            "line_item_id": 466157049,
+            "variant_id": 43729076,
+            "inventory_item_id": 808950810,
+            "inventory_commitment_id": null,
```
Comment (justinkrol) [inline test/fixtures/fulfillment_order.json]: Still missing an `inventory_commitment_id` here. Since these fixtures seem to be completely isolated, I think the foreign key id values can be anything.

Comment (karmakaze) [reply test/fixtures/fulfillment_order.json]: I thought that `inventory_commitment_id` were nullable and should be filled in when it made sense to do so rather than always having a value to clear out when not needed. Is that not the case?

File: test/fulfillment_order_test.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        new_location_id = 5
+
+        moved = ActiveSupport::JSON.decode(load_fixture('fulfillment_order'))
+        moved['fulfillment_order']['assigned_location_id'] = new_location_id
+        fake "orders/450789469/fulfillment_orders/519788021/move", :method => :post, :body => ActiveSupport::JSON.encode(moved)
+
+        assert_equal 'open', fulfillment_order.status
+        assert fulfillment_order.move(new_location_id: new_location_id)
+        assert_equal 'open', fulfillment_order.status
+        assert_equal new_location_id, fulfillment_order.assigned_location_id
```
Comment (justinkrol) [inline test/fulfillment_order_test.rb]: I think this assertion should actually check that `fulfillment_order` is closed.
When a FO is moved, we close the original and open a new one at the new location.

File: test/fixtures/fulfillment_order.json (hunk 2 above)
Comment (justinkrol) [reply test/fixtures/fulfillment_order.json]: You're right that a fulfillment order line item may not always have a corresponding commitment (like when a variant is deleted, maybe digital products in the future), but that's more of an edge case.

File: test/fulfillment_order_test.rb (hunk 3 above)
Comment (karmakaze) [reply test/fulfillment_order_test.rb]: I don't know how to map the response to the resource object that makes the request. The response from fulfillment_orders_controller has three keys
```
{
original_fulfillment_order:
moved_fulfillment_order:
remaining_fulfillment_order:
}
```
What should the state of 'resource` be after `resource.move`?

**Comment by karmakaze** [comment]
Adding `fulfillment_request`/`cancellation_request` actions to `FulfillmentOrder` resource.
Tried adding methods for filtering FOs by `assigned_status`/`location_ids` but it didn't like having its resource path altered from /orders/#/fulfillment_orders to /orders/#/assigned_fulfillment_orders.

File: test/fulfillment_order_test.rb [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        )
+
+        assert_equal 1, fulfillment_orders.count
+        fulfillment_order = fulfillment_orders.first
+        assert_equal 519788021, fulfillment_order.id
+        assert_equal 450789469, fulfillment_order.order_id
+      end
+    end
+
+    # context "#with_assigned_status" do
```
Comment (caueguerra) [inline test/fulfillment_order_test.rb]: what is up with this commented out code?

**Comment by stevelounsbury** [comment]
Hey @karmakaze, are you still working on this one?

File: test/fulfillment_order_test.rb (hunk 4 above)
Comment (karmakaze) [reply test/fulfillment_order_test.rb]: I was experimenting with how to organize assigned_fulfillment_orders either as a resource type matching the implemented endpoints or as methods on a fulfillment_orders resource from the client's perspective.

This can be deleted put into an assigned_fulfillment_orders_test.

**Comment by karmakaze** [comment]
@stevelounsbury Sorry I haven't had time to get back to Fulfillment Orders work, just finishing up the last of the Priority-1 Pickup In-store issues.

**Comment by karmakaze** [comment]
This work has been completed on branches #633 #635 (#664) #637 (#665) #639 (#666).
